src/Reports/index.ts

```diff
@@ -0,0 +1,56 @@
+import { Entity } from '../Entity.js'
+import { Centrifuge } from '../Centrifuge.js'
+import { processors } from './processors/index.js'
+import { ReportFilter } from './types.js'
+import { ReportService } from './ReportService.js'
+import { PoolSnapshotFilter, poolSnapshotsPostProcess, poolSnapshotsQuery } from '../queries/poolSnapshots.js'
+import {
+  TrancheSnapshotFilter,
+  trancheSnapshotsPostProcess,
+  trancheSnapshotsQuery,
+} from '../queries/trancheSnapshots.js'
+
+export class Reports extends Entity {
+  constructor(
+    centrifuge: Centrifuge,
+    public poolId: string
+  ) {
+    super(centrifuge, ['reports', poolId])
+  }
+
+  // TODO: add direction parameter (horizontal or vertical)
+  async balanceSheet(filter?: ReportFilter) {
+    const dateFilter = {
+      timestamp: {
+        greaterThan: filter?.from,
+        lessThan: filter?.to,
+      },
+    }
+
+    const [poolSnapshots, trancheSnapshots] = await Promise.all([
+      this.poolSnapshots({
+        ...dateFilter,
+        poolId: { equalTo: this.poolId },
+      }),
+      this.trancheSnapshots({
+        ...dateFilter,
+        tranche: { poolId: { equalTo: this.poolId } },
+      }),
+    ])
+
+    return ReportService.generate(processors.balanceSheet, this.poolId, { poolSnapshots, trancheSnapshots }, filter)
```

Comment:
I think the ReportService does more harm than good. The `poolSnapshots`, `trancheSnapshots` queries are already cached, so it's really only memoizing the process function. If `poolSnapshots` or `trancheSnapshots` return new data, this wouldn't necessarily be reflected in the result, because the ReportService might have a cached result. I would just call the process function directly.

Reply:
Can you please elaborate how you think the ReportService does harm? It seems like you're mostly just addressing the caching ability?

The cache is only valid for 5 minutes and it auto cleared after, I can reduce the time. And you got it right, the process function is exactly the one I'm trying to avoid being called n times every time its requested. I think the cache becomes even more relevant as the data set grows. We could be processing hundreds of thousands of data points multiple times as the user clicks into a graph (and switches in between graphs)

From a UX perspective we shouldn't just be updating data in a table without the user noticing. What we should do is let the user know that fresh data is available so they can click to update when ready. Kind of like how when you're reviewing PR and someone pushes something new

Reply:
Maybe we could implement a strategy that would invalidate the cache if there's new data available from the subquery

Reply:
Yes that would be the way to go. You'd want a "useMemo" and not a "useQuery". However, neither is necessary, because once the body of this function gets wrapped in a `_query`, you can pass the cacheKey to that and the result will be cached. Then the process function won't be called more than once, when it receives the same filters and the data is still the same.

The UX idea is a nice one and we can definitely take that into account when we start working on hooks and integrating into the App, but that's a userland problem. The SDK just provides the queries that users can await or subscribe to. They can choose how to display it and how the UI responds to updates. 

We could, for example, have a hook that queues new data and waits for user interaction before returning the new data. 
```
const { data, hasFreshData, refresh } = useObservableWithRefresh(...)
```

Reply:
You haven't elaborated how the caching does more harm than good, would be great if you could provide some clarity

Reply:
> The UX idea is a nice one and we can definitely take that into account when we start working on hooks and integrating into the App, but that's a userland problem. The SDK just provides the queries that users can await or subscribe to. They can choose how to display it and how the UI responds to updates.

while it does sound like a hooks challenge, to me it's also a problem we want to solve in the SDK first, it will influence how we push data updates through to the end user of the SDK